### PR TITLE
fix(dashboard): don't show report modal for anonymous user

### PR DESCRIPTION
**Pull request by mayurnewase**
### SUMMARY

Currently dashboard and chart checks for anonymous user by just check `user` variable existance in store, but anonymous user can exists with that object and permissions allowed on public role without authenticating and user id.
So logic inside reports fails as it tries to encode payload with undefined user id,so check for anonymous user by explicitly checking for valid user id is present.

### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before, when visiting a public dashboard without authentication
image

After, it loads without report button
image

### TESTING INSTRUCTIONS

Setup public dashboard and get public url
Setup public user with permissions to that dashboard
Log out and visit the dashboard

### ADDITIONAL INFORMATION

- [x] Has associated issue: https://github.com/apache/superset/issues/16718
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

**Review by geido — APPROVED** [review]
LGTM! Thanks for the tests as well!!

**Comment by mayurnewase** [comment]
@geido any chance this getting merged soon?

**Comment by geido** [comment]
Hey @mayurnewase I am asking for another pair of eye before merging. One of the committers will check in soon.

**Comment by kgabryje** [comment]
Thanks for adding tests @mayurnewase, looks great! Can you please rebase on the latest master and fix conflicts? Should be good to go then

File: superset-frontend/src/dashboard/components/Header/Header.test.tsx
Comment (michael-s-molina) [inline superset-frontend/src/dashboard/components/Header/Header.test.tsx:493]: Can we change this to `queryByRole` and `not.toBeInTheDocument`?

**Review by villebro — APPROVED** [review]
LGTM - however there's a conflict. Can you rebase and check if you agree with the recommended change raised by @michael-s-molina? Would be nice to get this merged and out of the way.

**Comment by mayurnewase** [comment]
> LGTM - however there's a conflict. Can you rebase and check if you agree with the recommended change raised by @michael-s-molina? Would be nice to get this merged and out of the way.

yes will do today.

**Comment by mayurnewase** [comment]
Not sure why prettier fails, I didnt touch any python file.

**Comment by villebro** [comment]
> Not sure why prettier is failing, I didn't touch any python file.

@mayurnewase I think that workflow checks frontend lint, too (appears to be the `Header.test.tsx` file that isn't passing prettier checking)

**Comment by villebro** [comment]
This appears to be the problem: 
image

**Comment by mayurnewase** [comment]
@villebro @michael-s-molina addressed your comments and rebased.
1 more review please?

**Review by michael-s-molina — APPROVED** [review]
LGTM. Thanks for the contribution!

**Comment by isavans** [comment]
This is still an issue in version 1.4.0.
Does anyone have an idea what version will include the fix for this.

**Pull request merged**
